python/pyspark/ml/torch/distributor.py

```diff
@@ -0,0 +1,491 @@
+            An integer that determines how many different concurrent
+            tasks are allowed. We expect spark.task.gpus = 1 for GPU-enabled training. Default
+            should be 1; we don't want to invoke multiple cores/gpus without explicit mention.
+        local_mode : bool, optional
+            A boolean that determines whether we are using the driver
+            node for training. Default should be false; we don't want to invoke executors without
+            explicit mention.
+        use_gpu : bool, optional
+            A boolean that indicates whether or not we are doing training
+            on the GPU. Note that there are differences in how GPU-enabled code looks like and
+            how CPU-specific code looks like.
+        spark : Optional[SparkSession], optional
+            An optional parameter that allows users to pass in a custom SparkSession argument
+            with a custom conf, by default None
+
+        Raises
+        ------
+        ValueError
+            If any of the parameters are incorrect.
+        """
+        super().__init__(num_processes, local_mode, use_gpu, spark)
+        self.framework = framework
+        self.ssl_conf = "pytorch.spark.distributor.ignoreSsl"  # type: ignore
+        self._validate_input_params()
+
+    def _validate_input_params(self) -> None:
+        """Validates input params
+
+        Raises
+        ------
+        ValueError
+            Thrown when user fails to provide correct input params
+        """
+        super()._validate_input_params()
+        if self.framework not in self.available_frameworks:
+            raise ValueError(
+                f"{self.framework} is not a valid framework."
+                f"Available frameworks: {self.available_frameworks}"
+            )
+
+    def _run_local_training(
+        self,
+        framework_wrapper_fn: Optional[Callable],
+        train_fn: Union[Callable, str],
+        *args: Any,
+    ) -> Optional[Any]:
+        CUDA_VISIBLE_DEVICES = "CUDA_VISIBLE_DEVICES"
+        old_cuda_visible_devices = os.environ.get(CUDA_VISIBLE_DEVICES, "")
+        cuda_state_was_set = CUDA_VISIBLE_DEVICES in os.environ
+        if not framework_wrapper_fn:
+            raise RuntimeError("Unknown combination of parameters")
+        try:
+            if self.use_gpu:
+                if "gpu" not in self.sc.resources:
+                    raise ValueError('"gpu" was not found in the context resources.')
+                gpus_owned = get_gpus_owned(self.sc.resources["gpu"].addresses)
+
+                if self.num_processes > len(gpus_owned):
+                    raise ValueError(
+                        f"""{self.num_processes} processes were requested
+                        for local training with GPU training but only
+                        {len(gpus_owned)} GPUs were available."""
+                    )
+                selected_gpus = [str(e) for e in random.sample(gpus_owned, self.num_processes)]
```

Comment:
Q: Why randomly pick GPUs ?

Reply:
If multiple runs happen concurrently, less chance that they will impact the same GPUs.

Reply:
But this may cause more flaky issue if the user rerun the function.

Reply:
So should we choose just the first `self.num_processes` GPUs owned then? I'm not sure what a good alternative approach here would be...

Reply:
@lu-wang-dl 

We can fix the randomizer seed.
e.g., we can generate the randomizer seed by the hash of the torch program code, 
then using the randomizer, we generate the selected_gpus list.

Reply:
Sure. It is fine if we add a dterministic random seed.